### ISSUE TITLE
Treesitter syntax and folding not immediately loading when Vim is launched like `vim path/to/file.lua`

### ISSUE THREAD

**Issue by asgeo1** [nvim-treesitter/nvim-treesitter#1469]
If I edit a file like this `vim path/to/file.lua` the treesitter highlighting, folding etc is not working. Treesitter seems to be loaded though, as I can use `:TSPlaygroundToggle` and see the normal tree output.

Running `:e %`, and treesitter syntax & folds start working fine.

However, if I open vim _first_, and then `:e /path/to/file.lua` it works fine from the start. Other plugins for opening files are also OK.  (except for Telescope, which fails on folding due to [this bug](https://github.com/nvim-telescope/telescope.nvim/issues/699))

**To Reproduce**

1. Run command `vim path/to/file.lua`
2. Notice syntax highlighting is not using treesitter.
3. `:e path/to/file.lua`
4. Notice syntax highlighting is now using treesitter.

**Expected behavior**

Launching Vim with `vim path/to/file.lua` should trigger syntax highlighting to work.

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>

health#nvim_treesitter#check
========================================================================
## Installation
  - OK: `tree-sitter` found  0.20.0 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v12.14.1 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl" }
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I
  - ruby           ✓ ✓ ✓ ✓ ✓ 
  - css            ✓ . ✓ ✓ ✓ 
  - cpp            ✓ ✓ ✓ ✓ ✓ 
  - swift          . . . . . 
  - json           ✓ ✓ ✓ ✓ . 
  - r              ✓ ✓ . . . 
  - python         ✓ ✓ ✓ ✓ ✓ 
  - vue            ✓ . ✓ . ✓ 
  - php            ✓ ✓ ✓ ✓ ✓ 
  - yaml           ✓ ✓ ✓ ✓ ✓ 
  - c              ✓ ✓ ✓ ✓ ✓ 
  - fennel         ✓ ✓ . . ✓ 
  - teal           ✓ ✓ ✓ ✓ ✓ 
  - ql             ✓ ✓ . ✓ ✓ 
  - verilog        ✓ ✓ ✓ . ✓ 
  - gomod          ✓ . . . . 
  - query          ✓ ✓ ✓ ✓ ✓ 
  - comment        ✓ . . . . 
  - jsdoc          ✓ . . . . 
  - sparql         ✓ ✓ ✓ ✓ ✓ 
  - gdscript       ✓ ✓ . . ✓ 
  - turtle         ✓ ✓ ✓ ✓ ✓ 
  - kotlin         ✓ . . . ✓ 
  - devicetree     ✓ ✓ ✓ ✓ ✓ 
  - julia          ✓ ✓ ✓ . ✓ 
  - regex          ✓ . . . . 
  - beancount      ✓ . ✓ . . 
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - latex          ✓ . ✓ . . 
  - elixir         ✓ ✓ . ✓ ✓ 
  - bibtex         ✓ . ✓ ✓ . 
  - go             ✓ ✓ ✓ ✓ ✓ 
  - zig            ✓ ✓ ✓ ✓ ✓ 
  - commonlisp     ✓ ✓ ✓ . . 
  - ocaml_interface✓ ✓ ✓ . ✓ 
  - ocamllex       ✓ . . . ✓ 
  - lua            ✓ ✓ ✓ ✓ ✓ 
  - cuda           ✓ ✓ ✓ ✓ . 
  - c_sharp        ✓ . ✓ . ✓ 
  - rust           ✓ ✓ ✓ ✓ ✓ 
  - ledger         ✓ . ✓ ✓ ✓ 
  - fortran        ✓ . ✓ ✓ . 
  - supercollider  ✓ ✓ ✓ ✓ ✓ 
  - ocaml          ✓ ✓ ✓ . ✓ 
  - elm            . . . . . 
  - nix            ✓ ✓ ✓ . ✓ 
  - typescript     ✓ ✓ ✓ ✓ ✓ 
  - glimmer        ✓ . . . . 
  - html           ✓ ✓ ✓ ✓ ✓ 
  - scss           ✓ . . ✓ . 
  - tsx            ✓ ✓ ✓ ✓ ✓ 
  - graphql        ✓ . . ✓ ✓ 
  - dockerfile     ✓ . . . ✓ 
  - toml           ✓ ✓ ✓ ✓ ✓ 
  - fish           ✓ ✓ ✓ ✓ ✓ 
  - jsonc          ✓ ✓ ✓ ✓ ✓ 
  - scala          . . . . . 
  - java           ✓ ✓ . ✓ ✓ 
  - dart           ✓ ✓ . ✓ ✓ 
  - rst            ✓ ✓ . . ✓ 
  - bash           ✓ ✓ ✓ . ✓ 
  - clojure        ✓ ✓ ✓ . ✓ 
  - svelte         ✓ . ✓ ✓ ✓ 
  - erlang         . . . . . 

 Legend: H[ighlight], L[ocals], F[olds], I[ndents]
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}

</code>
</details>

**Output of `nvim --version`**

<details>
<code>

NVIM v0.5.0-dev+1471-g7d8202087
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: clang -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/tmp/neovim-20210702-14071-8uo5z9/build/config -I/tmp/neovim-20210702-14071-8uo5z9/src -I/usr/local/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/usr/local/opt/gettext/include -I/tmp/neovim-20210702-14071-8uo5z9/build/src/nvim/auto -I/tmp/neovim-20210702-14071-8uo5z9/build/include
Compiled by asgeo1@Adams-MBP

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/Cellar/neovim/HEAD-7d82020_2/share/nvim"

Run :checkhealth for more info

</code>
</details>

**Additional context**

N/A

**Comment by farbodsz** [comment]
If you disable treesitter-based folding, do things work as expected?

(I'm wondering if it's related to #1412)

**Comment by asgeo1** [comment]
Yes, disabling these settings works:

```
vim.wo.foldmethod = "expr"
vim.wo.foldexpr = "nvim_treesitter#foldexpr()"
```

And then opening a file via `vim path/to/file.lua`, the treesitter syntax highlighting works now.  (with `indent` folding).

**Comment by stsewd** [comment]
Hi, can you reproduce this with a minimal init.vim?

```vim
call plug#begin('~/.local/share/nvim/plugged')
Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}
call plug#end()

set foldmethod=expr
set foldexpr=nvim_treesitter#foldexpr()
lua <<EOF
require'nvim-treesitter.configs'.setup {
  ensure_installed = "all",
  highlight = {
    enable = true,
  },
}
EOF
```

`nvim -u init.vim test.yaml`

I'm not able to reproduce this, maybe a plugin is breaking it?

**Comment by asgeo1** [comment]
@stsewd , this is the minimal example I can come up with for the syntax highlighting not working:

```
vim.cmd [[syntax enable]]

vim.wo.foldmethod = "expr"
vim.wo.foldexpr = "nvim_treesitter#foldexpr()"

local packer = require'packer'

packer.startup(
    function()
        use {
            "nvim-treesitter/nvim-treesitter",
            run = ":TSUpdate"
        }
    end
)

require("nvim-treesitter.configs").setup {
  highlight = {
    enable = true
  }
}
```

If you comment out the line `vim.cmd [[syntax enable]]` then treesitter syntax loads fine.  OR if you comment out the "folding" lines (but keep `syntax enable`), then it also loads treesitter syntax fine.

I'm not sure yet what is causing the folding to not work, seems to work in this minimal example. Will keep trying to find an example where that is not working.

**Comment by stsewd** [comment]
This looks like a problem with your package manager, I'm not able to replicate this using vim's pack feature. What about using packer's config option?

```lua
packer.startup(
  function()
    use {
      "nvim-treesitter/nvim-treesitter",
      run = ":TSUpdate",
      config = function ()
        require("nvim-treesitter.configs").setup {
          highlight = {
            enable = true,
          }
        }
      end
    }
  end
)

```

**Comment by asgeo1** [comment]
You're right, it seems to only happen when using packer. I tried vim-plug, and it was OK, even when using `syntax enable`.

I've only just moved to Packer, and to be honest, it's a bit frustrating how it works. It's difficult to enable/disable plugins for debugging issues like this.  And the comment you made about putting the `setup` inside `config` - it _should_ work like that to my understanding, yet it doesn't. I have to put `setup` outside `config`/`startup`, or no highlighting will be enabled. I've had a bunch of other issues too.

I'll raise this with the Vim packer project, and see what they say...

**Comment by stsewd** [comment]
@asgeo1 if you find the correct way to use this plugin with packer we would appreciate if you could update this section https://github.com/nvim-treesitter/nvim-treesitter/wiki/Installation#packernvim

**Comment by rod-stuchi** [comment]
Hi @asgeo1 did you solve the problem with packer? this issue still happens https://github.com/nvim-treesitter/nvim-treesitter/issues/1412, I'm with the same problem too.

**Comment by asgeo1** [comment]
No, I haven't solved it yet unfortunately. Since I mostly open files after I start Vim, it hasn't been a huge issue for me. But I would like to find a solution though

**Comment by pyrho** [comment]
> Hi @asgeo1 did you solve the problem with packer? this issue still happens #1412, I'm with the same problem too.

Same issue too, I had an `after/ftplugin/typescript_treesitter.vim` file with `foldmethod=expr` and `foldexpr=nvim_treesitter#foldexpr()`, removing that solved my issue as suggested in https://github.com/nvim-treesitter/nvim-treesitter/issues/1469#issuecomment-873282887

**Comment by 04l3x** [comment]
I have the same problem, I don't know if it can be useful for anyone, but I have added an `autocmd` function at the end of my init.vim, and this works for me, I'm not sure if this is ok but it works....

```vim
function FoldConfig()
	set foldmethod=expr
	set foldexpr=nvim_treesitter#foldexpr()
endfunction

autocmd BufAdd,BufEnter,BufNew,BufNewFile,BufWinEnter * :call FoldConfig()
```

**Comment by bennypowers** [comment]
I've added @04l3x 's workaround to [the wiki](https://github.com/nvim-treesitter/nvim-treesitter/wiki/Installation#packernvim)

**Comment by dajog05** [comment]
I got the same issue when using the [minpac](https://github.com/k-takata/minpac) package manager. Since both, Packer and minpac, use the native _packages_ feature of Vim, I suspected the issue somewhere there.

Under `:h runtime-search-path` I could find the following:
> When Vim starts up, after processing your .vimrc, it scans all directories in
'packpath' for plugins under the "pack/*/start" directory, and all the plugins
are loaded.

So to my understanding the problem seems to be that the plugins get loaded **after**  processing .vimrc.

Since `:packloadall` allows us to load packages immediately the following lines in my init.vim resolved the issue for me:

```vim
packloadall   " use before any treesitter configuration

set foldmethod=expr
set foldexpr=nvim_treesitter#foldexpr() 
```
I'm by no means an expert on Vim, so let me know if I got something wrong here, but at least on my system it seems to work.

**Comment by dajog05** [comment]
I got the same issue when using the [minpac](https://github.com/k-takata/minpac) package manager. Since both, Packer and minpac, use the native _packages_ feature of Vim, I suspected the issue somewhere there.

Under `:h runtime-search-path` I could find the following:
> When Vim starts up, after processing your .vimrc, it scans all directories in
'packpath' for plugins under the "pack/*/start" directory, and all the plugins
are loaded.

So to my understanding the problem seems to be that the plugins get loaded **after**  processing .vimrc.

Since `:packloadall` allows us to load packages immediately the following lines in my init.vim resolved the issue for me:

```vim
packloadall   " use before any treesitter configuration

lua <<EOF
require('nvim-treesitter.configs').setup( {
  highlight = {
    enable = true,
    additional_vim_regex_highlighting = false,
  },
})
EOF

set foldmethod=expr
set foldexpr=nvim_treesitter#foldexpr() 
```
However, you'll get an error message when opening a file that has an existing swap file. For files without an existing swap file this method executes normally on my system. Maybe someone with more experience and a better understanding of the codebase can figure out why.

I'm by no means an expert on Vim, so let me know if I got something wrong here, but at least on my system it seems to work so far.
